wallet/src/key_chain/master_key_chain.rs

```diff
@@ -0,0 +1,155 @@
+use crypto::key::hdkd::derivable::Derivable;
+use std::sync::Arc;
+use storage::Backend;
+use wallet_storage::{StoreTxRo, StoreTxRw, WalletStorageRead, WalletStorageWrite};
+use wallet_types::{AccountId, RootKeyContent, RootKeyId};
+use zeroize::Zeroize;
+
+#[allow(dead_code)] // TODO remove
+pub struct MasterKeyChain {
+    /// The specific chain this KeyChain is based on, this will affect the address format
+    chain_config: Arc<ChainConfig>,
+    /// The master key of this key chain from where all the keys are derived from
+    // TODO implement encryption
+    root_key: ExtendedPrivateKey,
+}
+
+impl MasterKeyChain {
+    #[allow(dead_code)] // TODO remove
+    pub fn mnemonic_to_root_key(
+        mnemonic_str: &str,
+        passphrase: Option<&str>,
+    ) -> KeyChainResult<ExtendedPrivateKey> {
+        let mut mnemonic = bip39::Mnemonic::parse(mnemonic_str).map_err(KeyChainError::Bip39)?;
+        let mut seed = mnemonic.to_seed(passphrase.unwrap_or(""));
+        let root_key = ExtendedPrivateKey::new_master(&seed, DEFAULT_KEY_KIND)?;
+        // TODO(SECURITY) confirm that the mnemonic is erased on drop
+        mnemonic.zeroize();
+        seed.zeroize();
+        Ok(root_key)
+    }
+
+    #[allow(dead_code)] // TODO remove
+    pub fn new_from_mnemonic<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        db_tx: &mut StoreTxRw<B>,
+        mnemonic_str: &str,
+        passphrase: Option<&str>,
+    ) -> KeyChainResult<Self> {
+        let root_key = Self::mnemonic_to_root_key(mnemonic_str, passphrase)?;
+        Self::new_from_root_key(chain_config, db_tx, root_key)
+    }
+
+    #[allow(dead_code)] // TODO remove
+    pub fn new_from_root_key<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        db_tx: &mut StoreTxRw<B>,
+        root_key: ExtendedPrivateKey,
+    ) -> KeyChainResult<Self> {
+        if !root_key.get_derivation_path().is_root() {
+            return Err(KeyChainError::KeyNotRoot);
+        }
+
+        let key_id = RootKeyId::from(root_key.to_public_key());
+        let key_content = RootKeyContent::from(root_key);
+
+        db_tx.set_root_key(&key_id, &key_content)?;
+
+        let root_key = key_content.into_key();
+
+        Ok(MasterKeyChain {
+            chain_config,
+            root_key,
+        })
+    }
+
+    /// Load the Master key chain from database and all the account key chains it derives
+    #[allow(dead_code)] // TODO remove
+    pub fn load_from_database<B: Backend>(
+        chain_config: Arc<ChainConfig>,
+        db_tx: &StoreTxRo<B>,
+    ) -> KeyChainResult<Self> {
+        let mut root_keys = db_tx.get_all_root_keys()?;
+
+        // The current format supports a single root key
+        if root_keys.len() != 1 {
+            return Err(KeyChainError::OnlyOneRootKeyIsSupported);
+        }
+
+        let (_, key_content) =
+            root_keys.pop_first().expect("Should not fail because it contains 1 key/value");
+
+        let root_key = key_content.into_key();
```

Comment:
Nitpick: I think `Itertools::exactly_one` could make this part a bit more concise.